### PR TITLE
Verdant/replace boost optional vals

### PR DESCRIPTION
**Pull request by kartikarcot**
This PR will completely remove boost::optional usage from gtsam. 

* All `boost::optional` usage is not replaced with `std::optional`
* Look at `std_optional_serialization.h` which is a mini implementation for serializing `std::optional`
* The `testStdOptionalSerialization.cpp` has some simple tests for the above feature. Let me know if I should
   test on more cases.
* All `boost::none` are replaced with `{}`
* In `OptionalJacobian` class removed the constructors that accepted `boost::optional<Matrix&>`
* Another note-worthy diff is in `EliminateableFactorGraph.h` where a `boost::optional<const VariableIndex&>` was used.
```
    /// Typedef for an optional variable index as an argument to elimination functions
   -  typedef boost::optional<const VariableIndex&> OptionalVariableIndex;
   + typedef const VariableIndex* OptionalVariableIndex;
```

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -6,7 +6,11 @@ if(NOT DEFINED CMAKE_MACOSX_RPATH)
   set(CMAKE_MACOSX_RPATH 0)
 endif()
 
-set(CMAKE_CXX_STANDARD 17)
+option(GTSAM_NO_BOOST_CPP17 "Require and use boost" ON)
+
+if (GTSAM_NO_BOOST_CPP17)
```
Comment (dellaert) [inline CMakeLists.txt:11]: Sorry for the misunderstanding, but we will move to c++17 no matter what, so this ifdef is not needed. Also, I'd like to remove this flag completely - we switch to c++17 using the build flags.

File: CMakeLists.txt [hunk 2]
```diff
@@ -6,7 +6,11 @@ if(NOT DEFINED CMAKE_MACOSX_RPATH)
   set(CMAKE_MACOSX_RPATH 0)
 endif()
 
-set(CMAKE_CXX_STANDARD 17)
+option(GTSAM_NO_BOOST_CPP17 "Require and use boost" ON)
```
Comment (dellaert) [inline CMakeLists.txt:9]: Is this the right place? I think maybe this flag should be defined in `HandleBoost.cmake`?

File: CMakeLists.txt [hunk 3]
```diff
@@ -53,7 +57,11 @@ endif()
 include(cmake/HandleGeneralOptions.cmake)   # CMake build options
 
 # Libraries:
-include(cmake/HandleBoost.cmake)            # Boost
+if (GTSAM_NO_BOOST_CPP17)
```
Comment (dellaert) [inline CMakeLists.txt:60]: Same, let's bundle everything in HandleBoost.cmake

**Review by dellaert — COMMENTED** [review]
Quick comments about configuration while I scan over the remainder of this massive PR :-)

File: CMakeLists.txt (hunk 2 above)
Comment (kartikarcot) [reply CMakeLists.txt:9]: This diff should go away. I forgot to remove it

File: examples/TriangulationLOSTExample.cpp [hunk 4]
```diff
@@ -131,23 +132,23 @@ int main(int argc, char* argv[]) {
...
     auto dltOptStart = std::chrono::high_resolution_clock::now();
-    boost::optional<Point3> estimateDLTOpt = triangulatePoint3<Cal3_S2>(
+    auto estimateDLTOpt = triangulatePoint3<Cal3_S2>(
         cameras, noisyMeasurements, rank_tol, true, measurementNoise, false);
     durationDLTOpt += std::chrono::high_resolution_clock::now() - dltOptStart;
 
-    errorsLOST.row(i) = *estimateLOST - landmark;
-    errorsDLT.row(i) = *estimateDLT - landmark;
-    errorsDLTOpt.row(i) = *estimateDLTOpt - landmark;
+    errorsLOST.row(i) = estimateLOST - landmark;
```
Comment (dellaert) [inline examples/TriangulationLOSTExample.cpp:149]: I don't understand - we still have to de-reference std::optionals as well, no?

File: CMakeLists.txt (hunk 1 above)
Comment (kartikarcot) [reply CMakeLists.txt:11]: It will go away

File: gtsam/base/OptionalJacobian.h [hunk 5]
```diff
@@ -66,11 +61,18 @@ class OptionalJacobian {
 
 public:
 
-  /// Default constructor acts like boost::none
+  /// Default constructor
   OptionalJacobian() :
       map_(nullptr) {
   }
 
+  /// Default constructor with nullptr_t
```
Comment (dellaert) [inline gtsam/base/OptionalJacobian.h:69]: Why was this not needed before?

File: gtsam/base/OptionalJacobian.h [hunk 6]
```diff
@@ -118,24 +120,6 @@ class OptionalJacobian {
     }
   }
 
-#ifndef OPTIONALJACOBIAN_NOBOOST
-
-  /// Constructor with boost::none just makes empty
```
Comment (dellaert) [inline gtsam/base/OptionalJacobian.h:123]: Don't we want to keep the std::optional path?

File: gtsam/base/OptionalJacobian.h [hunk 7]
```diff
@@ -200,7 +184,7 @@ class OptionalJacobian<Eigen::Dynamic, Eigen::Dynamic> {
 
 public:
 
-  /// Default constructor acts like boost::none
+  /// Default constructor acts like
```
Comment (dellaert) [inline gtsam/base/OptionalJacobian.h:187]: duplicate?

**Review by dellaert — COMMENTED** [review]
We should probably chat about OptionalJacobian

**Review by dellaert — COMMENTED** [review]
We should probably chat about OptionalJacobian

File: examples/TriangulationLOSTExample.cpp (hunk 4 above)
Comment (kartikarcot) [reply examples/TriangulationLOSTExample.cpp:149]: triangulatePoint3 returns a Point3 object. I don't know why we used optionals there. It should up when I replaced all boost::optionals with `auto` so I fixed it.

File: gtsam/base/OptionalJacobian.h (hunk 5 above)
Comment (kartikarcot) [reply gtsam/base/OptionalJacobian.h:69]: Because in this PR I replaced all `OptionalJacobian(boost::none)` to `OptionalJacobian(nullptr)`.

File: gtsam/geometry/triangulation.h [hunk 8]
```diff
@@ -656,6 +659,10 @@ class TriangulationResult : public boost::optional<Point3> {
   bool outlier() const { return status == OUTLIER; }
   bool farPoint() const { return status == FAR_POINT; }
   bool behindCamera() const { return status == BEHIND_CAMERA; }
+  const gtsam::Point3& get() const {
```
Comment (dellaert) [inline gtsam/geometry/triangulation.h:662]: Why is that here? Seems to not have been needed before.

File: gtsam/hybrid/HybridGaussianFactorGraph.h [hunk 9]
```diff
@@ -85,7 +87,7 @@ struct EliminationTraits<HybridGaussianFactorGraph> {
   /// The default ordering generation function
   static Ordering DefaultOrderingFunc(
       const FactorGraphType& graph,
-      boost::optional<const VariableIndex&> variableIndex) {
+      const VariableIndex&) {
```
Comment (dellaert) [inline gtsam/hybrid/HybridGaussianFactorGraph.h:90]: Same comment: should be std::reference_wrapper. No idea why HybridOrdering would not use it if available.

File: gtsam/inference/EliminateableFactorGraph-inst.h [hunk 10]
```diff
@@ -36,7 +36,7 @@ namespace gtsam {
       // no Ordering is provided.  When removing optional from VariableIndex, create VariableIndex
       // before creating ordering.
       VariableIndex computedVariableIndex(asDerived());
-      return eliminateSequential(orderingType, function, computedVariableIndex);
+      return eliminateSequential(orderingType, function, &computedVariableIndex);
```
Comment (dellaert) [inline gtsam/inference/EliminateableFactorGraph-inst.h:39]: should be restored after using reference wrapper

File: gtsam/inference/EliminateableFactorGraph.h [hunk 11]
```diff
@@ -89,10 +90,10 @@ namespace gtsam {
     typedef std::function<EliminationResult(const FactorGraphType&, const Ordering&)> Eliminate;
 
     /// Typedef for an optional variable index as an argument to elimination functions
-    typedef boost::optional<const VariableIndex&> OptionalVariableIndex;
+    typedef const VariableIndex* OptionalVariableIndex;
```
Comment (dellaert) [inline gtsam/inference/EliminateableFactorGraph.h:93]: same, restore if possible with std::reference_wrapper

File: gtsam/inference/EliminateableFactorGraph.h [hunk 12]
```diff
@@ -111,13 +112,13 @@ namespace gtsam {
      *  \code
      *  VariableIndex varIndex(graph); // Build variable index
      *  Data data = otherFunctionUsingVariableIndex(graph, varIndex); // Other code that uses variable index
-     *  boost::shared_ptr<GaussianBayesNet> result = graph.eliminateSequential(EliminateQR, varIndex, boost::none);
+     *  boost::shared_ptr<GaussianBayesNet> result = graph.eliminateSequential(EliminateQR, varIndex, std::nullopt);
```
Comment (dellaert) [inline gtsam/inference/EliminateableFactorGraph.h:115]: Why nullopt here rather than {}

File: gtsam/inference/ISAM-inst.h [hunk 13]
```diff
@@ -49,7 +49,7 @@ void ISAM<BAYESTREE>::updateInternal(const FactorGraphType& newFactors,
       KeyVector(newFactorKeys.begin(), newFactorKeys.end()));
 
   // eliminate all factors (top, added, orphans) into a new Bayes tree
-  auto bayesTree = factors.eliminateMultifrontal(ordering, function, index);
+  auto bayesTree = factors.eliminateMultifrontal(ordering, function, &index);
```
Comment (dellaert) [inline gtsam/inference/ISAM-inst.h:52]: undo

File: gtsam/linear/GaussianFactorGraph.h [hunk 14]
```diff
@@ -57,8 +58,8 @@ namespace gtsam {
     /// The default ordering generation function
     static Ordering DefaultOrderingFunc(
         const FactorGraphType& graph,
-        boost::optional<const VariableIndex&> variableIndex) {
-      return Ordering::Colamd(*variableIndex);
+        const VariableIndex& variableIndex) {
```
Comment (dellaert) [inline gtsam/linear/GaussianFactorGraph.h:61]: reference wrapper

**Review by dellaert — COMMENTED** [review]
About halfway through, will first patch base PR

File: gtsam/linear/GaussianFactorGraph.h (hunk 14 above)
Comment (kartikarcot) [reply gtsam/linear/GaussianFactorGraph.h:61]: I was meaning to ask you this. Here it takes an optional but never checks the optional. That's why I replaced it with just a reference. Do you want to keep the optional functionaltiy?

Comment (dellaert) [reply gtsam/linear/GaussianFactorGraph.h:61]: Yes.

File: examples/TriangulationLOSTExample.cpp (hunk 4 above)
Comment (dellaert) [reply examples/TriangulationLOSTExample.cpp:149]: @akshay-krishnan should LOST triangulation also not return none when it fails?

File: gtsam/base/OptionalJacobian.h (hunk 5 above)
Comment (dellaert) [reply gtsam/base/OptionalJacobian.h:71]: I think nullptr was already accepted as an argument, but maybe it was silently cast to boost::none?

File: gtsam/symbolic/SymbolicFactorGraph.h [hunk 15]
```diff
@@ -49,8 +49,8 @@ namespace gtsam {
     /// The default ordering generation function
     static Ordering DefaultOrderingFunc(
         const FactorGraphType& graph,
-        boost::optional<const VariableIndex&> variableIndex) {
-      return Ordering::Colamd(*variableIndex);
+        const VariableIndex& variableIndex) {
```
Comment (dellaert) [inline gtsam/symbolic/SymbolicFactorGraph.h:52]: reference wrapper

File: gtsam/base/OptionalJacobian.h (hunk 7 above)
Comment (kartikarcot) [reply gtsam/base/OptionalJacobian.h]: fixed

File: gtsam/base/OptionalJacobian.h (hunk 5 above)
Comment (kartikarcot) [reply gtsam/base/OptionalJacobian.h:71]: Possible that's what happend

File: CMakeLists.txt [hunk 16]
```diff
@@ -6,6 +6,8 @@ if(NOT DEFINED CMAKE_MACOSX_RPATH)
   set(CMAKE_MACOSX_RPATH 0)
 endif()
 
+set(CMAKE_CXX_STANDARD 17)
```
[the remark this answers is not in the document]
Comment (kartikarcot) [reply CMakeLists.txt]: done

File: gtsam_unstable/slam/SmartRangeFactor.h [hunk 17]
```diff
@@ -105,8 +106,8 @@ class SmartRangeFactor: public NoiseModelFactor {
     }
 
     Circle2 circle1 = circles.front();
-    boost::optional<Point2> best_fh;
-    auto bestCircle2 = boost::make_optional(false, circle1);  // fixes issue #38
+    std::optional<Point2> best_fh;
+    std::optional<Circle2> bestCircle2 = std::nullopt;  // fixes issue #38
```
Comment (dellaert) [inline gtsam_unstable/slam/SmartRangeFactor.h:110]: Sure about this?

File: gtsam/base/tests/testOptionalJacobian.cpp [hunk 18]
```diff
@@ -32,7 +32,6 @@ using namespace gtsam;
 TEST( OptionalJacobian, Constructors ) {
   Matrix23 fixed;
   Matrix dynamic;
-  boost::optional<Matrix&> optional(dynamic);
```
[the remark this answers is not in the document]
Comment (kartikarcot) [reply gtsam/base/tests/testOptionalJacobian.cpp:35]: done

File: gtsam/base/tests/testOptionalJacobian.cpp [hunk 19]
```diff
@@ -41,22 +40,18 @@ TEST( OptionalJacobian, Constructors ) {
   TEST_CONSTRUCTOR(2, 3, &fixed, true);
   TEST_CONSTRUCTOR(2, 3, dynamic, true);
   TEST_CONSTRUCTOR(2, 3, &dynamic, true);
-  TEST_CONSTRUCTOR(2, 3, boost::none, false);
```
[the remark this answers is not in the document]
Comment (kartikarcot) [reply gtsam/base/tests/testOptionalJacobian.cpp:44]: done

File: gtsam/inference/EliminateableFactorGraph-inst.h (hunk 10 above)
Comment (kartikarcot) [reply gtsam/inference/EliminateableFactorGraph-inst.h]: done

File: gtsam/inference/EliminateableFactorGraph.h (hunk 11 above)
Comment (kartikarcot) [reply gtsam/inference/EliminateableFactorGraph.h]: done

File: gtsam/inference/ISAM-inst.h (hunk 13 above)
Comment (kartikarcot) [reply gtsam/inference/ISAM-inst.h]: done

File: tests/testSimulated3D.cpp [hunk 20]
```diff
@@ -46,7 +46,7 @@ TEST( simulated3D, Values )
 TEST( simulated3D, Dprior )
 {
   Point3 x(1,-9, 7);
-  Matrix numerical = numericalDerivative11<Point3, Point3>(std::bind(simulated3D::prior, std::placeholders::_1, boost::none),x);
+  Matrix numerical = numericalDerivative11<Point3, Point3>(std::bind(simulated3D::prior, std::placeholders::_1, nullptr),x);
```
Comment (dellaert) [inline tests/testSimulated3D.cpp:49]: {} ?

**Review by dellaert — APPROVED** [review]
A few more minor comments, questions. Amazing work!

File: tests/testSimulated3D.cpp (hunk 20 above)
Comment (kartikarcot) [reply tests/testSimulated3D.cpp:49]: std:::bind has issues inferring type

File: gtsam_unstable/slam/SmartRangeFactor.h (hunk 17 above)
Comment (kartikarcot) [reply gtsam_unstable/slam/SmartRangeFactor.h:110]:
```
template<class T> inline optional<T> make_optional ( bool condition, T const& v ) ;  [new in 1.34]

```
This signature will set the optional variable to `v` if `condition` is true. Here it is hardcoded to false. That's why I did this. Also there is no equivalent in std::make_optional.

File: gtsam/symbolic/SymbolicFactorGraph.h (hunk 15 above)
Comment (kartikarcot) [reply gtsam/symbolic/SymbolicFactorGraph.h]: done

File: gtsam/base/TestableAssertions.h [hunk 21]
```diff
@@ -63,22 +63,14 @@ bool assert_equal(const boost::optional<V>& expected,
...
+bool assert_equal(const V& expected, const std::optional<V>& actual, double tol = 1e-9) {
   if (!actual) {
-    std::cout << "actual is boost::none" << std::endl;
+    std::cout << "actual is {}" << std::endl;
     return false;
   }
   return assert_equal(expected, *actual, tol);
 }
 
-template<class V>
```
[the remark this answers is not in the document]
Comment (kartikarcot) [reply gtsam/base/TestableAssertions.h:74]: It has reference to an optional. I can bring it back with std::reference_wrappper

File: examples/TriangulationLOSTExample.cpp (hunk 4 above)
Comment (akshay-krishnan) [reply examples/TriangulationLOSTExample.cpp:149]: *Likely unrelated to this PR*

The DLT code throws a `TriangulationUnderconstrainedException`, but LOST does not. Is that what you mean? I dont see any other differences. 
This difference comes from which API we use to solve the linear system. 
DLT: https://github.com/borglab/gtsam/blob/c7e18e64e59b61d158318e469759420ca9c78f20/gtsam/geometry/triangulation.cpp#L85
LOST: https://github.com/borglab/gtsam/blob/c7e18e64e59b61d158318e469759420ca9c78f20/gtsam/geometry/triangulation.cpp#L124

File: gtsam/inference/EliminateableFactorGraph.h (hunk 12 above)
Comment (kartikarcot) [reply gtsam/inference/EliminateableFactorGraph.h:116]: I think I did that make the documentation clear

File: gtsam/nonlinear/ISAM2UpdateParams.h [hunk 22]
```diff
@@ -37,16 +37,16 @@ struct ISAM2UpdateParams {
 
   /** An optional map of keys to group labels, such that a variable can be
    * constrained to a particular grouping in the BayesTree */
-  boost::optional<FastMap<Key, int>> constrainedKeys{boost::none};
+  std::optional<FastMap<Key, int>> constrainedKeys{std::nullopt};
```
[the remark this answers is not in the document]
Comment (kartikarcot) [reply gtsam/nonlinear/ISAM2UpdateParams.h:40]: done

File: gtsam/base/TestableAssertions.h (hunk 21 above)
[the remark this answers is not in the document]
Comment (kartikarcot) [reply gtsam/base/TestableAssertions.h:74]: done

[the remark this answers is not in the document]
Comment (kartikarcot) [reply gtsam/base/TestableAssertions.h:74]: added a test which uses this feature in `testMatrix.cpp`

File: gtsam/geometry/tests/testSimpleCamera.cpp [hunk 23]
```diff
@@ -0,0 +1,160 @@
+/* ----------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (kartikarcot) [reply gtsam/geometry/tests/testSimpleCamera.cpp]: sorry about that. fixed it

File: gtsam/discrete/DiscreteFactorGraph.h [hunk 24]
```diff
@@ -70,8 +70,8 @@ template<> struct EliminationTraits<DiscreteFactorGraph>
   /// The default ordering generation function
   static Ordering DefaultOrderingFunc(
       const FactorGraphType& graph,
-      boost::optional<const VariableIndex&> variableIndex) {
-    return Ordering::Colamd(*variableIndex);
+      const VariableIndex& variableIndex) {
```
[the remark this answers is not in the document]
Comment (kartikarcot) [reply gtsam/discrete/DiscreteFactorGraph.h]: fixed

File: gtsam/geometry/triangulation.h (hunk 8 above)
Comment (kartikarcot) [reply gtsam/geometry/triangulation.h:662]: It inherits from std::optional. boost::optional had a `get` call. Now we inherit from `std::optional`. So need to implement it

**Comment by dellaert** [comment]
Strange it got closed. Maybe automatic re-targeting does not work for forks...
